ADAL/src/ADEnrollmentGateway.m

```diff
@@ -110,71 +116,114 @@ + (NSString *)allEnrollmentIds
     return [[NSUserDefaults standardUserDefaults] objectForKey:AD_INTUNE_ENROLLMENT_ID_KEY];
 }
 
-+ (NSString *)enrollmentIdForUserId:(NSString *)userId error:(NSError**) error
++ (NSString *)allIntuneMAMResourcesJSON
+{
+    @synchronized(self)
+    {
+        if (s_intuneResourceJSON)
+            return s_intuneResourceJSON;
+    }
+    
+    return [[NSUserDefaults standardUserDefaults] objectForKey:AD_INTUNE_RESOURCE_ID_KEY];
+}
+
++ (NSString *)enrollmentIdForUserId:(NSString *)userId error:(ADAuthenticationError*__autoreleasing *) error
 {
     return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * dic) {
         return [[dic objectForKey:USER_ID] isEqualToString:userId];
     }
                                                        error:error];
 }
 
-+ (NSString *)enrollmentIdForUserObjectId:(NSString *)userObjectId tenantId:(NSString *)tenantId error:(NSError**) error
++ (NSString *)enrollmentIdForUserObjectId:(NSString *)userObjectId tenantId:(NSString *)tenantId error:(ADAuthenticationError*__autoreleasing *) error
 {
     return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * dic) {
         return [[dic objectForKey:OID] isEqualToString:userObjectId] && [[dic objectForKey:TID] isEqualToString:tenantId];
     }
                                                        error:error];
 }
 
-+ (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId error:(NSError**) error
++ (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId error:(ADAuthenticationError*__autoreleasing *) error
 {
     return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * dic) {
         return [[dic objectForKey:UNIQUE_ACCOUNT_ID] isEqualToString:uniqueAccountId];
     }
                                                        error:error];
 }
 
-+ (NSString *)intuneMamResource:(NSString *)authority error:(NSError**) error
++ (NSString *)enrollmentIdIfAvailable:(ADAuthenticationError**) error
 {
-    NSString* resourceJSON = nil;
-
-    @synchronized(self)
-    {
-        if (s_intuneResourceJSON)
-        {
-            resourceJSON = s_intuneResourceJSON;
-        }
+    // this will just return the first enrollment ID
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * __unused dic) {
+        return true;
     }
+                                                       error:error];
+}
+
++ (NSString*)enrollmentIDForUniqueAccountID:(NSString*) homeUserID userID:(NSString*) userID error:(ADAuthenticationError*__autoreleasing *) error
+{
+    NSString* enrollmentID;
```

Comment:
I think this logic should be following (@jasoncoolmax, please double check):
`
if (homeAccountID)
{
    // If homeAccountID is provided, always require an exact match
    return [ADEnrollmentGateway enrollmentIdForUniqueAccountId:homeUserID error:error];
}
else
{
    NSString *enrollmentID = nil;

    // If legacy userID is provided and we didn't find an exact match, do a fallback to any enrollment ID to support no userID or single userID scenarios
    if (userID) enrollmentID = [ADEnrollmentGateway enrollmentIdForUserId:userID error:error];
    if (enrollmentID) return enrollmentID;

    return ADEnrollmentGateway enrollmentIdIfAvailable:error];
}

`

Basically, I don't think we need a fallback if homeAccountId was provided.

Reply:
I think the above solution makes sense to me. 

I want to call out that, after server side make the changes (returning us user info together with auth code), we will remove all fallback in this function and only have exact match :)